### ISSUE TITLE
BUG: "output" information leakage into "input_ids" fields - incorrect attention_mask setup

### ISSUE THREAD

**Issue by Nsigma-Bill** [tloen/alpaca-lora#532]
It shares the same issue as Stanford Alpaca: https://github.com/tatsu-lab/stanford_alpaca/issues/290

Inside the function `generate_and_tokenize_prompt` of finetune.py:
`tokenized_full_prompt['input_ids']` has information on `instruction`, `input`, and `output` which is not correct (it should only contain information on `instruction` and `input`), i.e., `attention_mask` is not correctly setup.

**Comment by Nsigma-Bill** [comment]
Updated link:
https://github.com/huggingface/transformers/blob/fd6735102abcc560cb2b68523b3f5012da54a956/src/transformers/models/llama/modeling_llama.py#L460

**Comment by cfeng16** [comment]
Also sort of confused. Any thoughts?

**Comment by mitchelldehaven** [comment]
@Nsigma-Bill , did you get some more understanding on this topic. I still don't see how the fact that the input_ids in the Alpaca dataset are not offset by 1 from the labels. Your point of data leakage from the input ids seems valid and I don't see how the attention mask addresses that. The input_ids at idx `i` for example is equal to the label ids at `i`, assuming `i` is not in the prompt portion of the input. This seems incorrect to me, but I've seen this implemented this way in multiple places. Can you explain why you closed the issue?

**Comment by Nsigma-Bill** [comment]
@cfeng16 @mitchelldehaven https://github.com/huggingface/transformers/blob/fd6735102abcc560cb2b68523b3f5012da54a956/src/transformers/models/llama/modeling_llama.py#L460
In my opinion, I think what Hugging Face did is that they post-process your input to make sure it satisfies causality (i.e., prepare a causal attention mask).

**Comment by mitchelldehaven** [comment]
@Nsigma-Bill Doesn't the casual mask only prevent the model from cheating and looking right to the current index? If the model is predicting at index `i`, it still has access to information from the token at `i`, however not `i+1`. The problem here is that `input_ids[i] == labels[i]`. Am I making sense, or am I wrong somewhere?
